### ISSUE TITLE
@lint {Credo.Check.Design.DuplicatedCode, false} doesn't work on ExUnit setup blocks

### ISSUE THREAD

**Issue by KronicDeth** [rrrene/credo#103]
Using `@lint {Credo.Check.Design.DuplicatedCode, false}` does not disable the warning for `setup` blocks for ExUnit like

```
  @lint {Credo.Check.Design.DuplicatedCode, false}
  setup do
    {:ok, channel} = AMQP.Channel.open(Connection.get)
    # purge queue to remove undelivered messages from prior failing/interrupted tests and to ensure server and client
    # declare queue
    AMQP.Queue.purge(channel, @queue)
  end
```

I may be using the `@lint` wrong and knowing that is fine, or `@lint` may not be expected to work on anything besides literal `def` functions, which is fine too I just need to know that.

**Comment by rrrene** [comment]
This came up multiple times during ElixirConf.EU. It unfortunately is the current behaviour, but we all want that to change :grin:

As in, it would be good if `@lint` worked on macros besides the `def`, `defp` and `defmacro` macros, which are the only ones currently supported.

Related issue: #73

**Comment by KronicDeth** [comment]
Is there any way to now or any plans to allow turning off a rule for an entire file?  I use credo on my `lib` `web`, and `test`, but for instance of `Credo.Check.Design.DuplicatedCode`, I'll want to disable it for tests (since it can't be disabled for `setup`), but I still want to call out duplication in `lib` and `web`

**Comment by rrrene** [comment]
I was thinking about supporting some kind of syntax in the checks config to control which rules run on which files, but then I decided against even tinkering since this would lead to a lot of unnecessary complexity in Credo.

I would much prefer making the idiomatic `@lint` work as expected.

**Comment by aaronjensen** [comment]
@rrrene eslint has a great way of handling this--subdirectories can have their own `.eslintrc` and it gets merged into the parent directory. This means we can ignore specific rules for test directories and the like. It'd be nice if credo had this as I'd like to be able to credo my test directory but there are too many things that I'd want to ignore. Ignoring them manually is tedious.

**Comment by KronicDeth** [comment]
I don't think a directory level config file is a good solution in Elixir.  Elixir has the Mix.Config syntax and the predefined location for files per environment.  Don't fight that convention.

**Comment by aaronjensen** [comment]
In that case then we could just have the credo.exs return a list of configurations. Since it's just elixir you can build your configurations as you want including/excluding what you want. It's unclear what it would do if two configurations matched--it may merge them or pick the first.

**Comment by rrrene** [comment]
@aaronjensen You can have most of what you ask by adding a special config to your `credo.exs`:

```elixir
    %{
      name: "my_test_config",
      files: %{
        included: ["test/"],
        excluded: []
      },
      requires: [],
      check_for_updates: true,
      checks: [
        {Credo.Check.Readability.FunctionNames},
        {Credo.Check.Readability.LargeNumbers},
        {Credo.Check.Readability.MaxLineLength},
        {Credo.Check.Readability.ModuleAttributeNames},
        {Credo.Check.Readability.ModuleDoc},
        {Credo.Check.Readability.ModuleNames},
        {Credo.Check.Readability.ParenthesesInCondition},
        {Credo.Check.Readability.PredicateFunctionNames},
        {Credo.Check.Readability.TrailingBlankLine},
        {Credo.Check.Readability.TrailingWhiteSpace},
        {Credo.Check.Readability.VariableNames},
      ]
    }
```

This would disable all Readability checks when you run `mix credo --config-name=my_test_config`.

**Comment by philippneugebauer** [comment]
Can someone confirm, that `@lint` also do not work for `test` functions?

**Comment by rrrene** [comment]
I am closing this, since the bug/behaviour/functionality described here is related to `@lint` attributes, which will be removed in favor of more flexible comment-based syntax.

Please refer to #291 for the related discussion :+1:
